### PR TITLE
E2224 Refactor review_mapping_controller

### PR DESCRIPTION
**Pull request by HariniBharata**
Refactoring review_mapping_controller. @namanshrimali 
The project aims to refactor a controller - review_mapping_controller. It is a long and complex file and most of the methods do not include comments. The goal is to better re-structure it without modifying its original functionality and improving readability by using meaningful variable names and adding more comments.

### CODE REVIEW

**Comment by namanshrimali** [comment]
Hi Team,
It looks like the [build for Travis](https://app.travis-ci.com/github/expertiza/expertiza/jobs/563957440) is failing on your pull request, and the reason for the failure is because you've about 6 test cases that are failing for review_mapping_controller_spec.rb file. I checked the tests present in the beta branch on my system and they seem to pass. This means your changes broke existing functionality. Please have a look at the failure log and make changes accordingly.

Thanks,
Naman Shrimali

File: app/controllers/review_mapping_controller.rb [hunk 1]
```diff
@@ -325,82 +331,103 @@ def delete_metareview
...
+                                        participants, teams, num_reviews_per_student = 0,
                                         submission_review_num = 0)
     participants_hash = {}
-    participants.each { |participant| participants_hash[participant.id] = 0 }
+    participants.each {|participant| participants_hash[participant.id] = 0 }
     # calculate reviewers for each team
-    if !student_review_num.zero? && submission_review_num.zero?
-      review_strategy = ReviewMappingHelper::StudentReviewStrategy.new(participants, teams, student_review_num)
-    elsif student_review_num.zero? && !submission_review_num.zero?
+    if num_reviews_per_student != 0 and submission_review_num == 0
```
Comment (namanshrimali) [inline app/controllers/review_mapping_controller.rb:413]: Please use `zero?` method wherever possible.

File: app/controllers/review_mapping_controller.rb [hunk 2]
```diff
@@ -325,82 +331,103 @@ def delete_metareview
...
+    teams_with_uncalibrated_artifacts = teams - teams_with_calibrated_artifacts
+    # REVIEW: mapping strategy
+    automatic_review_mapping_strategy(assignment_id, participants, teams_with_calibrated_artifacts.shuffle!,  number_of_calibrated_artifacts, 0)
+    # REVIEW: mapping strategy
+    # since after first mapping, participants (delete_at) will be nil
+    participants = AssignmentParticipant.where(parent_id: params[:id].to_i).to_a.select(&:can_review).shuffle!
+    automatic_review_mapping_strategy(assignment_id, participants, teams_with_uncalibrated_artifacts.shuffle!, number_of_uncalibrated_artifacts, 0)
+  end
+
   def automatic_review_mapping
```
Comment (namanshrimali) [inline app/controllers/review_mapping_controller.rb:384]: Can you break down this method into smaller methods? ?

File: app/controllers/review_mapping_controller.rb [hunk 3]
```diff
@@ -325,82 +331,103 @@ def delete_metareview
...
     flash[:error] = params[:msg] if params[:msg]
     @assignment = Assignment.find(params[:id])
-    # ACS Removed the if condition(and corresponding else) which differentiate assignments as team and individual assignments
+    # ACS Removed the if condition(and corressponding else) which differentiate assignments as team and individual assignments
     # to treat all assignments as team assignments
     @items = AssignmentTeam.where(parent_id: @assignment.id)
     @items.sort_by(&:name)
   end
 
+  def create_team(participants,assignment_id, teams)
```
Comment (namanshrimali) [inline app/controllers/review_mapping_controller.rb:340]: Can you break down this method into smaller methods? Please add some comments over what this method does

File: app/controllers/review_mapping_controller.rb [hunk 4]
```diff
@@ -455,16 +482,36 @@ def start_self_review
   private
 
   def assign_reviewers_for_team(assignment_id, review_strategy, participants_hash)
-    if ReviewResponseMap.where(reviewed_object_id: assignment_id, calibrate_to: 0)
-                        .where('created_at > :time',
-                               time: @@time_create_last_review_mapping_record).size < review_strategy.reviews_needed
-
+    if check_eligibility(assignment_id,review_strategy)
       participants_with_insufficient_review_num = []
       participants_hash.each do |participant_id, review_num|
```
Comment (namanshrimali) [inline app/controllers/review_mapping_controller.rb:487]: Can you check whether [map](https://www.geeksforgeeks.org/ruby-array-map-function/) method can be used in this case?

File: app/controllers/review_mapping_controller.rb [hunk 5]
```diff
@@ -473,105 +520,141 @@ def assign_reviewers_for_team(assignment_id, review_strategy, participants_hash)
...
+    # If there are some submitters or reviewers in this team, they are not treated as normal participants.
+    # They should be removed from 'num_participants_this_team'
+    TeamsUser.where(team_id: team.id).each do |team_user|
+      temp_participant = Participant.where(user_id: team_user.user_id, parent_id: assignment_id).first
+      num_participants_this_team -= 1 if temp_participant.can_review == false or temp_participant.can_submit == false
     end
+    return num_participants_this_team
+  end
+
+  def condition_for_else(participants_hash)
```
Comment (namanshrimali) [inline app/controllers/review_mapping_controller.rb:635]: Please refactor the method name. `condition_for_else` does not make any sense.

File: app/controllers/review_mapping_controller.rb [hunk 6]
```diff
@@ -473,105 +520,141 @@ def assign_reviewers_for_team(assignment_id, review_strategy, participants_hash)
...
+    participants_with_min_assigned_reviews = participants_with_min_assigned_reviews(participants_hash)
+    # if participants_with_min_assigned_reviews is blank
+    if_condition_1 = participants_with_min_assigned_reviews.empty?
```
Comment (namanshrimali) [inline app/controllers/review_mapping_controller.rb:638]: Please rename variables `if_condition_1` and `if_condition_2` to a more informative ones.

File: spec/controllers/review_mapping_controller_spec.rb [hunk 7]
```diff
@@ -523,11 +523,11 @@
           allow_any_instance_of(ReviewMappingController).to receive(:automatic_review_mapping_strategy).with(any_args).and_return(true)
           params = {
             id: 1,
-            max_team_size: 1,
```
Comment (namanshrimali) [inline spec/controllers/review_mapping_controller_spec.rb:526]: Was there any specific cause for these renamings?

File: app/controllers/review_mapping_controller.rb [hunk 8]
```diff
@@ -473,105 +520,141 @@ def assign_reviewers_for_team(assignment_id, review_strategy, participants_hash)
...
+        break
       end
     end
-    @@time_create_last_review_mapping_record = ReviewResponseMap
-                                               .where(reviewed_object_id: assignment_id)
-                                               .last.created_at
+    return teams_hash
+  end
+
+  def update_participants_after_assigning_reviews(num_participants, maximum_reviews_per_student)
```
Comment (namanshrimali) [inline app/controllers/review_mapping_controller.rb:545]: Can you rename this method to match the description?

File: app/controllers/review_mapping_controller.rb [hunk 9]
```diff
@@ -455,16 +482,36 @@ def start_self_review
...
+      #can be added in a method
+      teams_hash = generate_teams_hash(assignment_id)
+      #can be added in a method
+      teams_hash = teams_hash_modified(participants_with_insufficient_review_num, teams_hash, assignment_id)
+    end
+    @@time_create_last_review_mapping_record = last_created_time_review_mapping_record(assignment_id)
+  end
+
+  # This method is used to check eligibility 
+  def check_eligibility(assignment_id,review_strategy)
```
Comment (namanshrimali) [inline app/controllers/review_mapping_controller.rb:499]: What eligibility does this method check? Can you rename it to more informative one?

File: spec/controllers/review_mapping_controller_spec.rb (hunk 7 above)
Comment (SriPallaviDamuluri) [reply spec/controllers/review_mapping_controller_spec.rb:526]: @namanshrimali , To make them more readable and it was mentioned in the documentation as well

File: app/controllers/review_mapping_controller.rb [hunk 10]
```diff
@@ -22,25 +20,22 @@ def action_allowed?
           'assign_quiz_dynamically',
           'start_self_review'
       true
-    else current_user_has_instructor_privileges?
```
Comment (namanshrimali) [inline app/controllers/review_mapping_controller.rb:25]: Is there a reason why this code was removed?

File: app/controllers/review_mapping_controller.rb [hunk 11]
```diff
@@ -212,82 +245,127 @@ def add_metareviewer
...
     redirect_to controller: 'student_review', action: 'list', id: metareviewer.id
   end
 
+  # This method gets the reviewer form the AssignmentParticipant. 
+  # If this operation fails, the code handles it by throwing an error. 
+  # If the reviewer does not exist, an error is thrown, asking the user to
+  # register the user first. 
+  # @param user is {Participant} type.
+  # @param reg_url is {string} 
   def get_reviewer(user, assignment, reg_url)
```
Comment (namanshrimali) [inline app/controllers/review_mapping_controller.rb:270]: You're not supposed to use get or set methods. Please change it to `reviewer`

File: app/controllers/review_mapping_controller.rb [hunk 12]
```diff
@@ -212,82 +245,127 @@ def add_metareviewer
...
-    reviewer = AssignmentParticipant.where(user_id: user.id, parent_id: assignment.id).first
-    raise "\"#{user.name}\" is not a participant in the assignment. Please <a href='#{reg_url}'>register</a> this user to continue." if reviewer.nil?
-
-    reviewer.get_reviewer
-  rescue StandardError => e
-    flash[:error] = e.message
+    begin
+      reviewer = AssignmentParticipant.where(user_id: user.id, parent_id: assignment.id).first
```
Comment (namanshrimali) [inline app/controllers/review_mapping_controller.rb:272]: Placing AssignmentParticipant object in `reviewer` is misleading. Can you rename this variable name?

File: app/controllers/review_mapping_controller.rb (hunk 10 above)
Comment (sravanth1) [reply app/controllers/review_mapping_controller.rb]: Reverted this change, as it was incorrectly taken from master to beta.

File: app/controllers/review_mapping_controller.rb [hunk 13]
```diff
@@ -22,25 +20,22 @@ def action_allowed?
...
+    else ['Instructor', 'Teaching Assistant', 'Administrator'].include? current_role_name
     end
   end
 
+ 
+  # The method add_calibration checks if a participant with  certain user_id exists in the
+  # AssignmentParticipant map. If the participant doesn't exist, this method creates a new participant 
+  # in the AssignmentParticipant map. This method also checks the ReviewResponseMap with certain parameters. 
+  # If the record doesn't exist, then a new record is created in the map. 
   def add_calibration
```
[the remark this answers is not in the document]
Comment (sravanth1) [reply app/controllers/review_mapping_controller.rb]: The add_calibration was being used to allow instructor to provide an expert review for an assignment. Changed the naming accordingly.

**Pull request closed without merging**